virtual_labs/infrastructure/stripe/webhook.py

```diff
@@ -0,0 +1,611 @@
+        if current_period_end is not None:
+            subscription.current_period_end = datetime.fromtimestamp(
+                float(str(current_period_end))
+            )
+
+        subscription.cancel_at_period_end = stripe_subscription.get(
+            "cancel_at_period_end", False
+        )
+        canceled_at = stripe_subscription.get("canceled_at")
+        if canceled_at is not None:
+            subscription.canceled_at = datetime.fromtimestamp(float(str(canceled_at)))
+
+        ended_at = stripe_subscription.get("ended_at")
+        if ended_at is not None:
+            subscription.ended_at = datetime.fromtimestamp(float(str(ended_at)))
+
+        billing_cycle_anchor = stripe_subscription.get("billing_cycle_anchor")
+        if billing_cycle_anchor is not None:
+            subscription.billing_cycle_anchor = datetime.fromtimestamp(
+                float(str(billing_cycle_anchor))
+            )
+
+        latest_invoice = stripe_subscription.get("latest_invoice")
+        if latest_invoice is not None:
+            if isinstance(latest_invoice, dict):
+                subscription.latest_invoice = latest_invoice.get("id")
+            else:
+                subscription.latest_invoice = latest_invoice
+
+        default_payment_method = stripe_subscription.get("default_payment_method")
+        if default_payment_method is not None:
+            if isinstance(default_payment_method, dict):
+                subscription.default_payment_method = default_payment_method.get("id")
+            else:
+                subscription.default_payment_method = default_payment_method
+
+        if items_data:
+            item = items_data[0]
+            price = item.get("price", {})
+
+            subscription.stripe_price_id = price.get("id")
+            subscription.amount = price.get("unit_amount", 0)
+            subscription.currency = price.get("currency", "usd")
+
+            recurring = price.get("recurring")
+            if recurring:
+                subscription.interval = recurring.get("interval", "month")
+
+        if (
+            event_type == "customer.subscription.deleted"
+            or subscription.status != "active"
+        ):
+            # TODO: update the user custom property "plan" to "free" in KC
+            await self.subscription_repository.downgrade_to_free(
+                user_id=subscription.user_id
+            )
+
+        db_session.add(subscription)
+        await db_session.commit()
+        await db_session.refresh(subscription)
+
+        return subscription
+
+    async def _handle_payment_event(
+        self, event_json: Dict[str, Any], db_session: AsyncSession
+    ) -> Dict[str, Any]:
+        """
+        handle invoice payment events by updating payment records.
+        """
+        event_type: str = str(event_json.get("type"))
+        event_data: Dict[str, Any] = event_json.get("data", {}).get("object", {})
+
+        print("ᦨ #  webhook.py:162 #  event_data:", event_data)
```

Comment:
I guess these are debug statements, I've seen such in other files as well. Would be nice to clean up at some point.